### PR TITLE
Fixes for common json errors, cleanup json_parser file.

### PR DESCRIPTION
**Pull request by BillSchumacher**
### Background

This fixes most errors I've encountered. This is from debugging the vicuna branch.

### Changes

Implemented serveral method to correct invalid JSON.

### Test Plan

Tested locally on vicuna branch.

### Change Safety

- [ ] I have added tests to cover my changes
- [x] I have considered potential risks and mitigations for my changes

### CODE REVIEW

**Comment by BillSchumacher** [comment]
I'm going to merge this back into my local model branch to ensure I didn't break anything with latest master changes, hold off on testing.

**Comment by BillSchumacher** [comment]
Lastest master changes kinda nuked what was working, I'm not really able to test this well without making a ton of openai calls.

**Review by Torantulino — APPROVED** [review]

**Comment by Torantulino** [comment]
Tested, seems to be working great.
Great work!

File: scripts/json_parser.py
Comment (DixonDs) [inline scripts/json_parser.py:37]: `correct_json` tries to parse the message as json in `balance_braces` so if it fails, we never get to the logic below. In my case it failed because the GPT returned a text before valid json, something like "Here's a command that you can use..." Was `correct_json` meant to be included in `try...except` below?

**Comment by DixonDs** [comment]
I appreciate all the efforts of the contributors and maintainers of the repo, but could it be a good idea to slow down and start requiring at least some unit tests? It looks like Wild West now, every time I pull changes, something is fixed but something else is broken :)

**Comment by ylluminate** [comment]
Still also seeing this frequently after a very recent pull:
```
SYSTEM:  Command Error: threw the following error: Invalid JSON
```

**Comment by ghzgod** [comment]
> I appreciate all the efforts of the contributors and maintainers of the repo, but could it be a good idea to slow down and start requiring at least some unit tests? It looks like Wild West now, every time I pull changes, something is fixed but something else is broken :)
I was just talking to my friend about this, I was like "I wonder if they are unit testing because things are breaking" haha.  I agree, we should slow it down to increase the reliability and the user base.

**Comment by hanskraut31312312** [comment]
> Still also seeing this frequently after a very recent pull:
> 
> ```
> SYSTEM:  Command Error: threw the following error: Invalid JSON
> ```

Me to SYSTEM:  Command Error: threw the following error: Invalid JSON loop be careful it just runs continuously even if you have not autorized not sure if it does use OpenAI API tokens tho. :)

File: scripts/json_parser.py
Comment (BillSchumacher) [reply scripts/json_parser.py:37]:
```  
      if balanced_str := balance_braces(json_str):
            return balanced_str
```

If the result is None it returns the original json, this is caught by a try/except.

All operations in the correct json are protected with try statements.

Comment (DixonDs) [reply scripts/json_parser.py:37]: In `balance_braces` itself you try to parse json and just rethrow an error (which is pointless tbh):

```
try:
	json.loads(json_string)
	return json_string
except json.JSONDecodeError as e:
	raise e
```

It is not caught in `correct_json `, it is used in the `except` part without any exception handling

Comment (BillSchumacher) [reply scripts/json_parser.py:37]: Yeah, you're right actually. You should submit a PR.

Comment (BillSchumacher) [reply scripts/json_parser.py:37]: Good catch

**Comment by fabricehong** [comment]
> 

You're right. We could also use gitflow in order to have a stable master branch.
-> https://github.com/Torantulino/Auto-GPT/discussions/728

**Comment by fabricehong** [comment]
> You're right. We could also use gitflow in order to have a stable master branch.
-> https://github.com/Torantulino/Auto-GPT/discussions/728

**Comment by fabricehong** [comment]
> I was just talking to my friend about this, I was like "I wonder if they are unit testing because things are breaking" haha. I agree, we should slow it down to increase the reliability and the user base.

You're right. We could also use gitflow in order to have a stable master branch.
-> https://github.com/Torantulino/Auto-GPT/discussions/728

File: scripts/json_parser.py
Comment (BillSchumacher) [reply scripts/json_parser.py:37]: Resolved by #747

**Pull request merged**
